aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp

```diff
@@ -0,0 +1,206 @@
+#include <ATen/ATen.h>
+#include <c10/util/irange.h>
+#include <cmath>
+#include <tuple>
+#include <vector>
+
+#ifdef USE_FBGEMM
+#include <fbgemm/QuantUtils.h>
+#endif
+#include <ATen/native/quantized/cpu/quant_utils.h>
+
+namespace {
+void calculate_moving_average(
+    const at::Tensor& x,
+    const at::Tensor& averaging_const,
+    at::Tensor& running_min,
+    at::Tensor& running_max,
+    bool per_row_fake_quant,
+    int ch_axis) {
+  at::Tensor x_min, x_max;
+  if (per_row_fake_quant) {
+    TORCH_CHECK(
+        ch_axis == 0,
+        "Per-channel FakeQuant in fused_moving_avg_obs_fake_quant is only supported on axis == 0");
+    std::tie(x_min, x_max) = at::_aminmax(x, 1);
+  } else {
+    std::tie(x_min, x_max) = at::_aminmax(x);
+  }
+  const float* min_curr_val = x_min.data_ptr<float>();
+  const float* max_curr_val = x_max.data_ptr<float>();
+  // Moving Average Min/Max observer for input tensor
+  float* running_min_val = running_min.data_ptr<float>();
+  float* running_max_val = running_max.data_ptr<float>();
+  float* averaging_const_val = averaging_const.data_ptr<float>();
+  for (const auto i : c10::irange(x_min.numel())) {
+    running_min_val[i] = std::isinf(running_min_val[i]) ? min_curr_val[i]
+                                                        : running_min_val[i] +
+            *averaging_const_val * (min_curr_val[i] - running_min_val[i]);
+    running_max_val[i] = std::isinf(running_max_val[i]) ? max_curr_val[i]
+                                                        : running_max_val[i] +
+            *averaging_const_val * (max_curr_val[i] - running_max_val[i]);
+  }
+
+  return;
+}
+
+std::tuple<at::Tensor, at::Tensor> choose_qparams_fake_quant(
+    const at::Tensor& x,
+    const at::Tensor& inp_running_min,
+    const at::Tensor& inp_running_max,
+    at::Tensor& scale,
+    at::Tensor& zero_point,
+    bool per_row_fake_quant,
+    bool symmetric_quant,
+    int qmin,
+    int qmax,
+    int ch_axis) {
+  std::tuple<at::Tensor, at::Tensor> fake_quant_out;
+  at::Tensor x_min, x_max;
+  if (per_row_fake_quant) {
+    float* x_min_data = inp_running_min.data_ptr<float>();
+    float* x_max_data = inp_running_max.data_ptr<float>();
+
+    for (const auto i : c10::irange(inp_running_min.numel())) {
+#ifdef USE_FBGEMM
+      fbgemm::TensorQuantizationParams x_qparams{};
+      x_qparams = fbgemm::ChooseQuantizationParams(
+          x_min_data[i],
+          x_max_data[i],
+          qmin,
+          qmax,
+          symmetric_quant,
+          /* preserve sparsity */ // Set preserve_sparsity to true for
+                                  // symmetric quantization.
+          false /* force power of two*/);
```

Comment:
optional nit: `//` for comments, to allow people to use `/* ... */` to comment out chunks of code?